### ISSUE TITLE
cmd/go: test -bench implies "chatty"

### ISSUE THREAD

**Issue by tamird** [golang/go#18010]
Consider the following invocations:
```
go test github.com/cockroachdb/cockroach/pkg/util/stop
```
```
go test github.com/cockroachdb/cockroach/pkg/util/stop -bench -
```
the two should ostensibly produce the same result, but the latter does not omit the output of successful tests as the former does. This does not appear to be documented anywhere.

A somewhat related issue was previously filed (https://github.com/golang/go/issues/10713) and closed with https://github.com/golang/go/commit/e9827f62018b87482e647dd88aecdedb24c94680, even though the behaviour appears to be global depending on the value of the `-bench` flag rather than based on the type of receiver of `Log` and `Logf`.

**Comment by josharian** [comment]
Possibly related to #17603?

**Comment by tamird** [comment]
Doesn't seem related to me.

**Comment by minux** [comment]
Could you reproduce this on any of the standard packages?

And which version of Go are you using?

**Comment by tamird** [comment]
What do you mean by "the standard packages"? What does this have to do with the package being tested?

`go version go1.7.1 darwin/amd64`

**Comment by minux** [comment]
Because I can't reproduce this in any of the std packages that I tried,
so I suspect that maybe it has something to do with the package
being tested.

Are you suggesting that difference in output is the extra "PASS"
at the beginning of the output? If that's case, it's working as intended
as if -bench is passed, even if none of the benchmark matches,
the testing package will still use the output format intended for
benchmarks.

**Comment by mvdan** [comment]
@minux I think he means stdout/stderr produced by tests, which is hidden by `go test`.

**Comment by mvdan** [comment]
Also, what is the reason for the inconsistency between `-chatty` and `-v`/`-test.v` in the docs?

**Comment by minux** [comment]
I still don't understand what this issue is about.

$ cat a_test.go
package pkg

import (
"fmt"
"testing"
)

func TestA(t *testing.T) {
fmt.Println("Tstdout")
println("Tstderr")
t.Log("Tlog")
}

func BenchmarkA(t *testing.B) {
fmt.Println("Bstdout")
println("Bstderr")
t.Log("Blog")
}

$ go171 test
Tstdout
Tstderr
PASS
ok   pkg 0.001s
$ go171 test -bench -
Tstdout
Tstderr
PASS
ok   pkg 0.001s

**Comment by tamird** [comment]
@minux your example is exactly the problem. if you run `go test` (without `-bench -`) you will not see "Tstdout" nor "Tstderr".

**Comment by mvdan** [comment]
@tamird but he did just there and it also prints.

**Comment by tamird** [comment]
Oh, it seems the behaviour is not so simple.
```
$ (cd foo && go test)
Tstdout
Tstderr
PASS
ok  	github.com/cockroachdb/cockroach/foo	0.005s
$ (cd foo && go^Cest)
go test ./foo
ok  	github.com/cockroachdb/cockroach/foo	0.006s
$ go test ./foo -bench -
Tstdout
Tstderr
PASS
ok  	github.com/cockroachdb/cockroach/foo	0.006s
```
...so if you specify a package name, stdout/stderr are swallowed, but if you don't, they are not. this is so inconsistent so as to be crazy.

**Comment by minux** [comment]
OK, here are all four combinations:
$ go171 test pkg
ok   pkg 0.001s
$ go171 test pkg -bench -
Tstdout
Tstderr
PASS
ok   pkg 0.001s
$ go171 test
Tstdout
Tstderr
PASS
ok   pkg 0.001s
$ go171 test -bench -
Tstdout
Tstderr
PASS
ok   pkg 0.002s

I think this is all working as intended.

When -bench is not given,
1. with package name(s), go test only shows test summary
2. without package name, go test assume you're developing that package,
and want full output, so it doesn't suppress stdout/stderr.

When -bench is given, go test will always show stdout/stderr, even if it
doesn't match any of the benchmarks (this is because the benchmark
statistics is produced by the testing package, not by the go command,
so if the Go command wants to show you the benchmark result, it can't
hide the other output from the test.)

**Comment by tamird** [comment]
Working as intended by whom? None of this is documented, and none of it makes sense.

Why does supplying a package name suppress stdout? "go test assume you're developing that package" this is a rationalization, or something, because it's certainly not documented.

> 
When -bench is given, go test will always show stdout/stderr, even if it
doesn't match any of the benchmarks (this is because the benchmark
statistics is produced by the testing package, not by the go command,
so if the Go command wants to show you the benchmark result, it can't
hide the other output from the test.)

That sounds like an implementation detail, not a sensible reason that for this behaviour. Again, it is not documented.

**Comment by minux** [comment]
undocumented, perhaps, but it has always been this way since Go 1
so it's very unlikely to change.

**Comment by tamird** [comment]
why? it doesn't seem to me that any of this is covered by the Go 1 guarantee.

**Comment by minux** [comment]
What do you want to change?

Tools are not covered by Go 1 guarantee, but we should at least
preserve existing behavior as much as possible (in case people
have existing programs that parse such output).

**Comment by tamird** [comment]
As the issue description states, I propose that the `-bench` flag not affect stdout/stderr suppression.

**Comment by knz** [comment]
+1 I think the behavior currently in the code was actually not intended by the designer of this code, and I think anyone using the benchmarking code will *welcome* fixing this issue that has been there since Go 1.

**Comment by rsc** [comment]
Like @minux wrote above, this is working as intended (I designed and wrote the code, and this is what I intended). If running benchmarks, we stream the output from the benchmarks to make sure they are seen at all but also that they are seen as they happen.

The only option we've come up with is to make go test -bench run the binary twice, once for tests (quietly) and once for benchmarks (not quietly). That seems a bit odd, and a different special case (init code would run multiple times, and so on).
